CSV-Datei wird aufgemacht,die Daten werden weiter als DataFrame von Pandas betrachtet

In [1]:
import numpy as np
import pandas as pd

db = pd.read_csv("train_ver2.csv", sep=",")

/Users/dariagolchenko/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (5,8,11,15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Typumwandlung in DateType

In [2]:
print("fecha_dato -> " + str(db["fecha_dato"].isnull().sum()))
print("fecha_alta -> " + str(db["fecha_alta"].isnull().sum()))

fecha_dato -> 0
fecha_alta -> 27734


In [3]:
db["fecha_dato"]=pd.to_datetime(db["fecha_dato"], format = "%Y-%m-%d")
print(db["fecha_dato"].dtypes)

datetime64[ns]


Es gibt höhe Wahrscheinlichkeit, dass die Saisonaleschwankungen den höchen Einfluss auf die Nachfrage haben

In [4]:
db['month_of_transaction'] = db['fecha_dato'].dt.month
db.head()

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  fecha_alta  \
0 2015-01-28   1375586            N              ES    H   35  2015-01-12   
1 2015-01-28   1050611            N              ES    V   23  2012-08-10   
2 2015-01-28   1050612            N              ES    V   23  2012-08-10   
3 2015-01-28   1050613            N              ES    H   22  2012-08-10   
4 2015-01-28   1050614            N              ES    V   23  2012-08-10   

   ind_nuevo antiguedad  indrel          ...          ind_plan_fin_ult1  \
0        0.0          6     1.0          ...                          0   
1        0.0         35     1.0          ...                          0   
2        0.0         35     1.0          ...                          0   
3        0.0         35     1.0          ...                          0   
4        0.0         35     1.0          ...                          0   

  ind_pres_fin_ult1 ind_reca_fin_ult1 ind_tjcr_fin_ult1 ind_valo_fin_ult1  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

  ind_viv_fin_ult1 ind_nomina_ult1 ind_nom_pens_ult1  ind_recibo_ult1  \
0                0             0.0               0.0                0   
1                0             0.0               0.0                0   
2                0             0.0               0.0                0   
3                0             0.0               0.0                0   
4                0             0.0               0.0                0   

   month_of_transaction  
0                     1  
1                     1  
2                     1  
3                     1  
4                     1  

[5 rows x 49 columns]

Typumwandlung in geeigneten Typen; Erzeugeng der neuen Spalten mir kodierten Werten (es ist nötig für die richtige Berechnung der Korrelation)

In [5]:
db.isnull().any()

fecha_dato               False
ncodpers                 False
ind_empleado              True
pais_residencia           True
sexo                      True
age                      False
fecha_alta                True
ind_nuevo                 True
antiguedad               False
indrel                    True
ult_fec_cli_1t            True
indrel_1mes               True
tiprel_1mes               True
indresi                   True
indext                    True
conyuemp                  True
canal_entrada             True
indfall                   True
tipodom                   True
cod_prov                  True
nomprov                   True
ind_actividad_cliente     True
renta                     True
segmento                  True
ind_ahor_fin_ult1        False
ind_aval_fin_ult1        False
ind_cco_fin_ult1         False
ind_cder_fin_ult1        False
ind_cno_fin_ult1         False
ind_ctju_fin_ult1        False
ind_ctma_fin_ult1        False
ind_ctop_fin_ult1        False
ind_ctpp

In [6]:
print("age -> " + str(db["age"].isnull().sum()))
print("fecha_alta -> " + str(db["fecha_alta"].isnull().sum()))
print("ind_nuevo -> " + str(db["ind_nuevo"].isnull().sum()))
print("indrel -> " + str(db["indrel"].isnull().sum()))
print("conyuemp -> " + str(db["conyuemp"].isnull().sum()))
print("tipodom -> " + str(db["tipodom"].isnull().sum()))
print("ult_fec_cli_1t -> " + str(db["ult_fec_cli_1t"].isnull().sum()))
print("ind_actividad_cliente -> " + str(db["ind_actividad_cliente"].isnull().sum()))
print("renta -> " + str(db["renta"].isnull().sum()))
print("ind_nomina_ult1 -> " + str(db["ind_nomina_ult1"].isnull().sum()))
print("ind_nom_pens_ult1 -> " + str(db["ind_nom_pens_ult1"].isnull().sum()))
print("nomprov -> " + str(db["nomprov"].isnull().sum()))
print("count of rows -> " + str(db.shape))

age -> 0
fecha_alta -> 27734
ind_nuevo -> 27734
indrel -> 27734
conyuemp -> 13645501
tipodom -> 27735
ult_fec_cli_1t -> 13622516
ind_actividad_cliente -> 27734
renta -> 2794375
ind_nomina_ult1 -> 16063
ind_nom_pens_ult1 -> 16063
nomprov -> 93591
count of rows -> (13647309, 49)


In [7]:
print(str(db.loc[db.age.isnull(), "ncodpers"].unique()))
print(str(db.loc[db.fecha_alta.isnull(), "ncodpers"].unique()))
print(str(db.loc[db.ind_nuevo.isnull(), "ncodpers"].unique()))
print(str(db.loc[db.indrel.isnull(), "ncodpers"].unique()))
print(str(db.loc[db.tipodom.isnull(), "ncodpers"].unique()))
print(str(db.loc[db.ind_actividad_cliente.isnull(), "ncodpers"].unique()))
print(str(db.loc[db.age.isnull(), "ncodpers"].unique()))

[]
[1050741 1051017 1051064 ... 1125581 1401509  429713]
[1050741 1051017 1051064 ... 1125581 1401509  429713]
[1050741 1051017 1051064 ... 1125581 1401509  429713]
[1050741 1051017 1051064 ... 1125581 1401509  429713]
[1050741 1051017 1051064 ... 1125581 1401509  429713]
[]


In [8]:
print(str(db.loc[db.age.isnull(), "fecha_dato"].unique()))
print(str(db.loc[db.fecha_alta.isnull(), "fecha_dato"].unique()))
print(str(db.loc[db.ind_nuevo.isnull(), "fecha_dato"].unique()))
print(str(db.loc[db.indrel.isnull(), "fecha_dato"].unique()))
print(str(db.loc[db.tipodom.isnull(), "fecha_dato"].unique()))
print(str(db.loc[db.ind_actividad_cliente.isnull(), "fecha_dato"].unique()))
print(str(db.loc[db.age.isnull(), "fecha_dato"].unique()))

[]
['2015-01-28T00:00:00.000000000' '2015-02-28T00:00:00.000000000'
 '2015-03-28T00:00:00.000000000' '2015-04-28T00:00:00.000000000'
 '2015-05-28T00:00:00.000000000' '2015-06-28T00:00:00.000000000']
['2015-01-28T00:00:00.000000000' '2015-02-28T00:00:00.000000000'
 '2015-03-28T00:00:00.000000000' '2015-04-28T00:00:00.000000000'
 '2015-05-28T00:00:00.000000000' '2015-06-28T00:00:00.000000000']
['2015-01-28T00:00:00.000000000' '2015-02-28T00:00:00.000000000'
 '2015-03-28T00:00:00.000000000' '2015-04-28T00:00:00.000000000'
 '2015-05-28T00:00:00.000000000' '2015-06-28T00:00:00.000000000']
['2015-01-28T00:00:00.000000000' '2015-02-28T00:00:00.000000000'
 '2015-03-28T00:00:00.000000000' '2015-04-28T00:00:00.000000000'
 '2015-05-28T00:00:00.000000000' '2015-06-28T00:00:00.000000000'
 '2015-11-28T00:00:00.000000000']
['2015-01-28T00:00:00.000000000' '2015-02-28T00:00:00.000000000'
 '2015-03-28T00:00:00.000000000' '2015-04-28T00:00:00.000000000'
 '2015-05-28T00:00:00.000000000' '2015-06-28T00:00

In [5]:
db = db[db.fecha_alta.notnull()]

In [10]:
print("age -> " + str(db["age"].isnull().sum()))
print("fecha_alta -> " + str(db["fecha_alta"].isnull().sum()))
print("ind_nuevo -> " + str(db["ind_nuevo"].isnull().sum()))
print("indrel -> " + str(db["indrel"].isnull().sum()))
print("conyuemp -> " + str(db["conyuemp"].isnull().sum()))
print("tipodom -> " + str(db["tipodom"].isnull().sum()))
print("ult_fec_cli_1t -> " + str(db["ult_fec_cli_1t"].isnull().sum()))
print("ind_actividad_cliente -> " + str(db["ind_actividad_cliente"].isnull().sum()))
print("renta -> " + str(db["renta"].isnull().sum()))
print("ind_nomina_ult1 -> " + str(db["ind_nomina_ult1"].isnull().sum()))
print("ind_nom_pens_ult1 -> " + str(db["ind_nom_pens_ult1"].isnull().sum()))
print("nomprov -> " + str(db["nomprov"].isnull().sum()))
print("canal_entrada -> " + str(db["canal_entrada"].isnull().sum()))
print("count of rows -> " + str(db.shape))

age -> 0
fecha_alta -> 0
ind_nuevo -> 0
indrel -> 0
conyuemp -> 13617767
tipodom -> 1
ult_fec_cli_1t -> 13594782
ind_actividad_cliente -> 0
renta -> 2766641
ind_nomina_ult1 -> 217
ind_nom_pens_ult1 -> 217
nomprov -> 65857
canal_entrada -> 158392
count of rows -> (13619575, 49)


In [6]:
db = db[db.ind_nomina_ult1.notnull()]

In [12]:
print("conyuemp -> " + str(db["conyuemp"].isnull().sum()))
print("tipodom -> " + str(db["tipodom"].isnull().sum()))
print("ult_fec_cli_1t -> " + str(db["ult_fec_cli_1t"].isnull().sum()))
print("renta -> " + str(db["renta"].isnull().sum()))
print("nomprov -> " + str(db["nomprov"].isnull().sum()))
print("count of rows -> " + str(db.shape))

conyuemp -> 13617552
tipodom -> 1
ult_fec_cli_1t -> 13594566
renta -> 2766601
nomprov -> 65857
count of rows -> (13619358, 49)


In [7]:
db = db[db.nomprov.notnull()]

In [14]:
print("conyuemp -> " + str(db["conyuemp"].isnull().sum()))
print("tipodom -> " + str(db["tipodom"].isnull().sum()))
print("ult_fec_cli_1t -> " + str(db["ult_fec_cli_1t"].isnull().sum()))
print("renta -> " + str(db["renta"].isnull().sum()))
print("nomprov -> " + str(db["nomprov"].isnull().sum()))
print("canal_entrada -> " + str(db["canal_entrada"].isnull().sum()))
print("count of rows -> " + str(db.shape))

conyuemp -> 13551695
tipodom -> 0
ult_fec_cli_1t -> 13528746
renta -> 2700952
nomprov -> 0
canal_entrada -> 158090
count of rows -> (13553501, 49)


Gucken, welche Verteilung die Werte in der Spalte "age" haben

Korrelationskoeffizient (0,3 bis 0,5 ist schon eine mittlere Korrelation, > als 0,5 - schon starke Korrelation) bei:
- ind_cno_fin_ult1 und ind_reca_fin_ult1 ist 0.319811
- ind_actividad_cliente und ind_reca_fin_ult1 ist 0.222989
- ind_actividad_cliente und ind_tjcr_fin_ult1 ist 0.232320
- ind_recibo_ult1 und ind_reca_fin_ult1 ist 0.343852
- ind_recibo_ult1 und ind_tjcr_fin_ult1 ist 0.374660
- ind_cno_fin_ult1 und ind_nomina_ult1 ist 0.757957
- ind_nom_pens_ult1 und ind_nomina_ult1 ist 0.957215
- ind_recibo_ult1 und ind_nomina_ult1 ist 0.471138
- ind_actividad_cliente und ind_recibo_ult1 ist 0.412218
- ind_cno_fin_ult1 und ind_nom_pens_ult1 ist 0.790128
- nomprov_cat und cod_prov ist 0.855563

In [15]:
print(db.corr())


                       ncodpers  ind_nuevo    indrel  tipodom  cod_prov  \
ncodpers               1.000000   0.347391  0.028579      NaN -0.040823   
ind_nuevo              0.347391   1.000000  0.063041      NaN -0.004380   
indrel                 0.028579   0.063041  1.000000      NaN -0.000413   
tipodom                     NaN        NaN       NaN      NaN       NaN   
cod_prov              -0.040823  -0.004380 -0.000413      NaN  1.000000   
ind_actividad_cliente -0.140125  -0.004028 -0.033039      NaN  0.028168   
renta                 -0.084208  -0.014133  0.000022      NaN -0.015241   
ind_ahor_fin_ult1     -0.013120  -0.002560 -0.000434      NaN  0.001366   
ind_aval_fin_ult1     -0.005492  -0.000986 -0.000201      NaN  0.000525   
ind_cco_fin_ult1       0.162532   0.003123 -0.046986      NaN -0.005365   
ind_cder_fin_ult1     -0.018850  -0.004608 -0.000850      NaN  0.000881   
ind_cno_fin_ult1      -0.131195  -0.047304 -0.011587      NaN  0.012171   
ind_ctju_fin_ult1     -0.

Zu große Abweichung:

In [15]:
meanOfRenta = np.mean(db.renta)
medianOfRenta = np.nanmedian(db.renta)
print(meanOfRenta, medianOfRenta)

134252.63614951386 101848.77


Ich habe vermutet, dass die Höche der Einkommen ziemlich gut durch
    1. Geschlecht
    2. Arbeitsverhältnisse
    3. Region
erklärt werden können, prüfen:

In [8]:
renta_median_H = db.loc[db.sexo == "H","renta"].median()
renta_mean_H = db.loc[db.sexo == "H","renta"].mean()
renta_median_V = db.loc[db.sexo == "V","renta"].median()
renta_mean_V = db.loc[db.sexo == "V","renta"].mean()
print(str(renta_median_H))
print(str(renta_mean_H))
print(str(renta_median_V))
print(str(renta_mean_V))

99465.03
132330.99848332693
103806.15
135839.65725299172


In [9]:
db.ind_empleado.unique()

array(['N', 'A', 'F', 'B', 'S'], dtype=object)

In [10]:
renta_median_N = db.loc[db.ind_empleado == "N","renta"].median()
renta_mean_N = db.loc[db.ind_empleado == "N","renta"].mean()
renta_median_B = db.loc[db.ind_empleado == "B","renta"].median()
renta_mean_B = db.loc[db.ind_empleado == "B","renta"].mean()
renta_median_F = db.loc[db.ind_empleado == "F","renta"].median()
renta_mean_F = db.loc[db.ind_empleado == "F","renta"].mean()
renta_median_S = db.loc[db.ind_empleado == "S","renta"].median()
renta_mean_S = db.loc[db.ind_empleado == "S","renta"].mean()
print(str(renta_median_N))
print(str(renta_mean_N))
print(str(renta_median_B))
print(str(renta_mean_B))
print(str(renta_median_F))
print(str(renta_mean_F))
print(str(renta_median_S))
print(str(renta_mean_S))

101810.28
134218.9170161395
171413.91
204292.14416017485
149037.51
167212.1001293664
87798.24
87798.24


In [8]:
set = db.nomprov.unique()

In [20]:
for i in set:
    print(str(i) + " : " + str(db.loc[db.nomprov == i,"renta"].median()))
    print(str(i) + " : " + str(db.loc[db.nomprov == i,"renta"].mean()))

MALAGA : 95102.61
MALAGA : 121216.57669297257
CIUDAD REAL : 62193.45
CIUDAD REAL : 69888.62509656424
ZARAGOZA : 99950.28
ZARAGOZA : 110587.50481891104
TOLEDO : 68867.76
TOLEDO : 80624.57113916425
LEON : 80901.54000000001
LEON : 93383.86710125268
GIPUZKOA : 80599.14
GIPUZKOA : 141895.85610389616
CACERES : 67799.49
CACERES : 75365.83711824188
GIRONA : 108963.21
GIRONA : 144281.0302492002
ZAMORA : 74692.70999999999
ZAMORA : 83322.11130582851
BARCELONA : 131214.48
BARCELONA : 164679.42156618612
SALAMANCA : 89831.25
SALAMANCA : 105810.76795913748
BURGOS : 89372.34
BURGOS : 97880.72286751622
HUESCA : 73467.81
HUESCA : 89228.51724407451
NAVARRA : 86649.27
NAVARRA : 102476.17862068968
AVILA : 68561.28
AVILA : 76853.64482514081
SEGOVIA : 89311.56
SEGOVIA : 98463.74209361993
LUGO : 64390.53
LUGO : 76686.76213098167
LERIDA : 64818.45
LERIDA : 81212.51591857828
MADRID : 139070.97
MADRID : 178898.88464661813
ALICANTE : 67526.28
ALICANTE : 87357.72662546462
SORIA : 78810.9
SORIA : 88064.10784519262


In [9]:
for i in set:
    db.loc[(db.nomprov == str(i)) & (db.renta.isnull()),"renta"] = db.loc[db.nomprov == str(i),"renta"].median()
db.renta.isnull().sum()

0

In [23]:
pd.value_counts(db["canal_entrada"])

KHE    4050971
KAT    3242337
KFC    3089469
KHQ     591039
KFA     399290
KHK     241023
KHM     183859
KHN     116588
KHD     116000
KAS      85364
RED      74941
KAG      73516
KAY      66959
KAA      65229
KAB      62064
KAE      50250
KCC      48017
KBZ      45555
KHL      45102
KFD      43691
KAI      37597
KEY      35095
KAW      34020
KAZ      32118
KAR      31927
KAF      30317
007      28750
013      27095
KCI      25708
KAH      24705
        ...   
004        210
KCQ        198
KDN        197
KCR        196
KDH        191
KCO        179
KEE        175
KCP        158
K00        152
KBE        147
KEQ        146
KFB        107
KAV        107
KCT        105
KBX        102
KBP         85
KEM         68
KCX         67
KFV         67
KHA         51
KBN         44
KGU         28
KGC         28
KGN         17
KDI         17
KDB         17
KDL         11
025         11
KHS          5
KHR          1
Name: canal_entrada, Length: 162, dtype: int64

In [10]:
db.loc[db.canal_entrada.isnull(), "canal_entrada"] = "KHE"

In [11]:
db["sexo"] = db["sexo"].astype(str)
db.loc[db.sexo == "nan", "sexo"].count()

70

In [16]:
pd.value_counts(db["sexo"])

V      7382171
H      6171260
nan         70
Name: sexo, dtype: int64

In [12]:
db.loc[db.sexo == "nan", "sexo"] = "V"
db.loc[db.sexo == "nan", "sexo"].count()

0

In [30]:
db.isnull().any()

fecha_dato               False
ncodpers                 False
ind_empleado             False
pais_residencia          False
sexo                     False
age                      False
fecha_alta               False
ind_nuevo                False
antiguedad               False
indrel                   False
ult_fec_cli_1t            True
indrel_1mes               True
tiprel_1mes               True
indresi                  False
indext                   False
conyuemp                  True
canal_entrada            False
indfall                  False
tipodom                  False
cod_prov                 False
nomprov                  False
ind_actividad_cliente    False
renta                    False
segmento                  True
ind_ahor_fin_ult1        False
ind_aval_fin_ult1        False
ind_cco_fin_ult1         False
ind_cder_fin_ult1        False
ind_cno_fin_ult1         False
ind_ctju_fin_ult1        False
ind_ctma_fin_ult1        False
ind_ctop_fin_ult1        False
ind_ctpp

In [13]:
db["ind_empleado"] = db["ind_empleado"].astype(str)
db["pais_residencia"] = db["pais_residencia"].astype(str)
db["sexo"] = db["sexo"].astype("category")
db["indrel_1mes"] = db["indrel_1mes"].astype(str)
db["indresi"] = db["indresi"].astype(str)
db["indext"] = db["indext"].astype(str)
db["canal_entrada"] = db["canal_entrada"].astype(str)
db["indfall"] = db["indfall"].astype(str)
db["nomprov"] = db["nomprov"].astype(str)
db["segmento"] = db["segmento"].astype(str)
db["age"] = pd.to_numeric(db["age"], errors="coerce")
db["ult_fec_cli_1t"] = db["ult_fec_cli_1t"].astype(str)
db["conyuemp"] = pd.to_numeric(db["conyuemp"], errors="coerce")
db["ind_nomina_ult1"] = pd.to_numeric(db["ind_nomina_ult1"], errors="coerce")
db["ind_nom_pens_ult1"] = pd.to_numeric(db["ind_nom_pens_ult1"], errors="coerce")
db["renta"] = pd.to_numeric(db["renta"], errors="coerce")
db["fecha_alta"]=pd.to_datetime(db["fecha_alta"], format = "%Y-%m-%d")

Prüfen wir nochmal: (ich habe nur wichtige gelassen)

In [19]:
db.loc[db.indrel_1mes == "nan", ["indrel_1mes"]].count()

indrel_1mes    122038
dtype: int64

In [20]:
db.loc[db.segmento == "nan", ["segmento"]].count()

segmento    161319
dtype: int64

D.h. Es bleiben noch: indrel_1mes, segmento, tiprel_1mes, conyuemp, ult_fec_cli_1t

In [21]:
pd.value_counts(db["tiprel_1mes"])

I    7273215
A    6152751
P       4624
R        869
N          4
Name: tiprel_1mes, dtype: int64

In [22]:
db.tiprel_1mes.isnull().sum()

122038

In [14]:
db.loc[db.tiprel_1mes.isnull(), "tiprel_1mes"] = "I"
db.tiprel_1mes.isnull().any()

False

In [24]:
pd.value_counts(db["indrel_1mes"])

1.0    11235776
1       2188874
nan      122038
3.0        3183
3          1135
P           873
2.0         833
2           483
4.0         236
4            70
Name: indrel_1mes, dtype: int64

P - Potentional wird jetzt 5 (einheitlicher Datentyp)

In [15]:
db.loc[db.indrel_1mes == 1.0, ["indrel_1mes"]] = 1
db.loc[db.indrel_1mes == 3.0, ["indrel_1mes"]] = 3
db.loc[db.indrel_1mes == "P", ["indrel_1mes"]] = 5
db.loc[db.indrel_1mes == 2.0, ["indrel_1mes"]] = 2
db.loc[db.indrel_1mes == 4.0, ["indrel_1mes"]] = 4
db.loc[db.indrel_1mes == "nan", ["indrel_1mes"]] = 1
db["indrel_1mes"] = pd.to_numeric(db["indrel_1mes"], errors="coerce")
pd.value_counts(db["indrel_1mes"])

1.0    13546688
3.0        4318
2.0        1316
5.0         873
4.0         306
Name: indrel_1mes, dtype: int64

In [26]:
pd.value_counts(db["segmento"])

02 - PARTICULARES     7902975
03 - UNIVERSITARIO    4930811
01 - TOP               558396
nan                    161319
Name: segmento, dtype: int64

In [16]:
db.loc[db.segmento == "nan", ["segmento"]] = "02 - PARTICULARES"
db.loc[db.segmento.isnull(), ["segmento"]] = "02 - PARTICULARES"

In [41]:
pd.value_counts(db["segmento"])

02 - PARTICULARES     8064294
03 - UNIVERSITARIO    4930811
01 - TOP               558396
Name: segmento, dtype: int64

In [17]:
db.loc[db.ult_fec_cli_1t == "nan", ["ult_fec_cli_1t"]] = "0"

Prüfen wo noch nan-Values oder Fehler vorhanden sind: (Antwort: Fehler bei antiquedad)

In [42]:
for i in db.columns.values.tolist():
    print(str(i) + ": " + str(pd.value_counts(db[str(i)])))

fecha_dato: 2016-05-28    927461
2016-04-28    924284
2016-03-28    921088
2016-02-28    916917
2016-01-28    912292
2015-12-28    908044
2015-11-28    902130
2015-10-28    888275
2015-09-28    861464
2015-08-28    839224
2015-07-28    825842
2015-06-28    626567
2015-05-28    624659
2015-04-28    622360
2015-03-28    620404
2015-02-28    617724
2015-01-28    614766
Name: fecha_dato, dtype: int64
ncodpers: 32752      17
1134383    17
592934     17
576542     17
560150     17
543758     17
551946     17
527366     17
1347479    17
1314695    17
1298303    17
1306491    17
1281911    17
1290099    17
1273707    17
1240923    17
1216343    17
1199951    17
1183559    17
1191747    17
1167167    17
650298     17
642110     17
683082     17
945354     17
1043706    17
1027314    17
986342     17
994530     17
969950     17
           ..
1490796     1
1551281     1
1553389     1
936625      1
1549443     1
603198      1
1551998     1
1413214     1
1551001     1
1553384     1
711905      1
14

nomprov: MADRID                    4409491
BARCELONA                 1275204
VALENCIA                   682298
SEVILLA                    605148
CORUÑA, A                  429321
MURCIA                     396752
MALAGA                     367018
ZARAGOZA                   342542
ALICANTE                   313395
CADIZ                      294682
PONTEVEDRA                 280024
ASTURIAS                   265745
VALLADOLID                 238257
PALMAS, LAS                235695
BADAJOZ                    192995
BIZKAIA                    185888
TOLEDO                     183065
GRANADA                    178559
SALAMANCA                  164237
CANTABRIA                  155698
CORDOBA                    144678
CACERES                    129912
BALEARS, ILLES             124933
HUELVA                     122281
CIUDAD REAL                119389
ALBACETE                   114128
CASTELLON                  104294
TARRAGONA                  100113
BURGOS                      97186
GIRON

In [18]:
db.loc[db.antiguedad == -999999, ["antiguedad"]].count()

antiguedad    26
dtype: int64

In [19]:
db.loc[db.antiguedad == -999999, ["antiguedad"]] = 0

In [45]:
db.loc[db.antiguedad == "-999999", ["antiguedad"]].count()

antiguedad    12
dtype: int64

In [20]:
db.loc[db.antiguedad == "-999999", ["antiguedad"]] = 0

In [21]:
db.loc[db.ult_fec_cli_1t.isnull(), ["ult_fec_cli_1t"]] = "0"
db.ult_fec_cli_1t.isnull().sum()

0

In [22]:
print(pd.value_counts(db["conyuemp"]))
del db["conyuemp"]
print("count of rows -> " + str(db.shape))

Series([], Name: conyuemp, dtype: int64)
count of rows -> (13553501, 48)


In [23]:
db.isnull().any()

fecha_dato               False
ncodpers                 False
ind_empleado             False
pais_residencia          False
sexo                     False
age                      False
fecha_alta               False
ind_nuevo                False
antiguedad               False
indrel                   False
ult_fec_cli_1t           False
indrel_1mes              False
tiprel_1mes              False
indresi                  False
indext                   False
canal_entrada            False
indfall                  False
tipodom                  False
cod_prov                 False
nomprov                  False
ind_actividad_cliente    False
renta                    False
segmento                 False
ind_ahor_fin_ult1        False
ind_aval_fin_ult1        False
ind_cco_fin_ult1         False
ind_cder_fin_ult1        False
ind_cno_fin_ult1         False
ind_ctju_fin_ult1        False
ind_ctma_fin_ult1        False
ind_ctop_fin_ult1        False
ind_ctpp_fin_ult1        False
ind_deco

In [24]:
db["ind_empleado"] = db["ind_empleado"].astype(str)
db["pais_residencia"] = db["pais_residencia"].astype(str)
db["sexo"] = db["sexo"].astype("category")
db["indrel_1mes"] = db["indrel_1mes"].astype(str)
db["indresi"] = db["indresi"].astype(str)
db["indext"] = db["indext"].astype(str)
db["canal_entrada"] = db["canal_entrada"].astype(str)
db["indfall"] = db["indfall"].astype(str)
db["nomprov"] = db["nomprov"].astype(str)
db["segmento"] = db["segmento"].astype(str)
db["age"] = pd.to_numeric(db["age"], errors="coerce")
db["ult_fec_cli_1t"] = db["ult_fec_cli_1t"].astype(str)
db["ind_nomina_ult1"] = pd.to_numeric(db["ind_nomina_ult1"], errors="coerce")
db["ind_nom_pens_ult1"] = pd.to_numeric(db["ind_nom_pens_ult1"], errors="coerce")
db["renta"] = pd.to_numeric(db["renta"], errors="coerce")
db["fecha_alta"]=pd.to_datetime(db["fecha_alta"], format = "%Y-%m-%d")

In [25]:
list_cat = ["ind_empleado", "pais_residencia", "sexo", "indrel_1mes", "tiprel_1mes", "indresi", "indext", "canal_entrada", "indfall", "nomprov", "segmento"]

In [26]:

db_dumm = pd.get_dummies(db, columns=list_cat)

In [27]:
db_dumm.head()

fecha_dato  ncodpers  age fecha_alta  ind_nuevo antiguedad  indrel  \
0 2015-01-28   1375586   35 2015-01-12        0.0          6     1.0   
1 2015-01-28   1050611   23 2012-08-10        0.0         35     1.0   
2 2015-01-28   1050612   23 2012-08-10        0.0         35     1.0   
3 2015-01-28   1050613   22 2012-08-10        0.0         35     1.0   
4 2015-01-28   1050614   23 2012-08-10        0.0         35     1.0   

  ult_fec_cli_1t  tipodom  cod_prov             ...               \
0              0      1.0      29.0             ...                
1              0      1.0      13.0             ...                
2              0      1.0      13.0             ...                
3              0      1.0      50.0             ...                
4              0      1.0      50.0             ...                

   nomprov_TARRAGONA  nomprov_TERUEL  nomprov_TOLEDO  nomprov_VALENCIA  \
0                  0               0               0                 0   
1                  0               0               0                 0   
2                  0               0               0                 0   
3                  0               0               0                 0   
4                  0               0               0                 0   

   nomprov_VALLADOLID  nomprov_ZAMORA  nomprov_ZARAGOZA  segmento_01 - TOP  \
0                   0               0                 0                  0   
1                   0               0                 0                  0   
2                   0               0                 0                  0   
3                   0               0                 1                  0   
4                   0               0                 1                  0   

   segmento_02 - PARTICULARES  segmento_03 - UNIVERSITARIO  
0                           1                            0  
1                           0                            1  
2                           0                            1  
3                           0                            1  
4                           0                            1  

[5 rows x 282 columns]

In [28]:
print("count of rows -> " + str(db_dumm.shape))

count of rows -> (13553501, 282)


Und speichern als CSV:

In [29]:
db_dumm.to_csv("new_data_wiinf.csv", sep='\t', encoding='utf-8')

Ausgangszustand:
- 13647309 Datensätze;
- 48 Spalten, dafon 23 Attribute (inkl. Kundennummer und Transaktionsdatum)

Nach der Vorbereitung:
- 13553501 Datensätze --> Verlust 93808 Datensätze
- 47 Spalten --> Verlust 1 Attribut

Nach der Dummykodierung:
- 282 Spalten